src/data-model-providers/codegen/EmberAttributeStorageImpl.cpp

```diff
@@ -0,0 +1,164 @@
+/**
+ *    Copyright (c) 2025 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+#include "lib/core/CHIPError.h"
+#include <data-model-providers/codegen/EmberAttributeStorageImpl.h>
+
+#include <app-common/zap-generated/attribute-type.h>
+#include <app/util/persistence/AttributePersistenceProvider.h>
+#include <lib/support/ScopedBuffer.h>
+
+namespace chip {
+namespace app {
+namespace Storage {
+namespace {
+
+// EmberAfAttributeType possible values are part of the unnamed enum in `attribute-type.h`,
+// of the form ZCL_*_ATTRIBUTE_TYPE
+EmberAfAttributeType attributeType(AttributeStorage::Buffer & buffer)
+{
+    switch (buffer.type())
+    {
+    case AttributeStorage::Buffer::Type::kPrimitive:
+        switch (buffer.size())
+        {
+        case 1:
+            return ZCL_INT8U_ATTRIBUTE_TYPE;
+        case 2:
+            return ZCL_INT16U_ATTRIBUTE_TYPE;
+        case 3:
+            return ZCL_INT24U_ATTRIBUTE_TYPE;
+        case 4:
+            return ZCL_INT32U_ATTRIBUTE_TYPE;
+        case 5:
+            return ZCL_INT40U_ATTRIBUTE_TYPE;
+        case 6:
+            return ZCL_INT48U_ATTRIBUTE_TYPE;
+        case 7:
+            return ZCL_INT56U_ATTRIBUTE_TYPE;
+        case 8:
+            return ZCL_INT64U_ATTRIBUTE_TYPE;
+        default:
+            // This one is a bit nonsense..
+            return ZCL_NO_DATA_ATTRIBUTE_TYPE;
+        }
+    case AttributeStorage::Buffer::Type::kMutableByteSpan:
+        // Generally ember persistence did not persist raw values.
+        return ZCL_STRUCT_ATTRIBUTE_TYPE;
+    case AttributeStorage::Buffer::Type::kStringOneByteLength:
+    case AttributeStorage::Buffer::Type::kBytesOneByteLength:
+        return ZCL_OCTET_STRING_ATTRIBUTE_TYPE;
```

Comment:
Please either document why we claim octet string for char strings, or just return the right types for char strings?